### PR TITLE
Benchmarks!

### PR DESCRIPTION
**Pull request by Luthaf**
This PR add a set of regression benchmarks to lumol, and move the comparison benchmarks to https://github.com/lumol-org/benchmarks. The benchmarks uses the [bencher](https://crates.io/crates/bencher) crate to run on stable. So you can just run `cargo bench` and everything works out of the box.

It also fixes two bugs for the molecular systems: aef34ba177a170deac39ab15dee3b4767722515f and e9d11106b0b234b22583ec13fc32a2fab9c99fbf.

Here are the benchmark results for my machine:

```
$ cargo bench 
   Compiling lumol v0.1.0 (file:///home/guillaume/perso/lumol/lumol)
    Finished release [optimized] target(s) in 2.67 secs
     Running target/release/deps/argon-0a23a4f2120755ff

running 3 tests
test energy ... bench:   2,292,825 ns/iter (+/- 155,674)
test forces ... bench:   2,387,518 ns/iter (+/- 205,636)
test virial ... bench:   2,968,812 ns/iter (+/- 197,603)

test result: ok. 0 passed; 0 failed; 0 ignored; 3 measured

     Running target/release/deps/lumol-170d004fa597deab

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running target/release/deps/nacl-49663082ea4bd521

running 6 tests
test energy_ewald ... bench:     641,087 ns/iter (+/- 20,833)
test energy_wolf  ... bench:     518,130 ns/iter (+/- 16,887)
test forces_ewald ... bench:  36,168,774 ns/iter (+/- 2,201,070)
test forces_wolf  ... bench:     637,741 ns/iter (+/- 10,857)
test virial_ewald ... bench:  92,395,825 ns/iter (+/- 5,644,890)
test virial_wolf  ... bench:     640,696 ns/iter (+/- 24,558)

test result: ok. 0 passed; 0 failed; 0 ignored; 6 measured

     Running target/release/deps/propane-4c7de46b7ab84597

running 3 tests
test energy ... bench:   1,205,487 ns/iter (+/- 30,358)
test forces ... bench:   1,366,635 ns/iter (+/- 28,467)
test virial ... bench:   1,460,978 ns/iter (+/- 31,803)

test result: ok. 0 passed; 0 failed; 0 ignored; 3 measured

     Running target/release/deps/water-f42b2d61b47127cc

running 6 tests
test energy_ewald ... bench:   1,054,666 ns/iter (+/- 10,310)
test energy_wolf  ... bench:     787,354 ns/iter (+/- 32,793)
test forces_ewald ... bench:  49,983,830 ns/iter (+/- 2,459,616)
test forces_wolf  ... bench:     962,400 ns/iter (+/- 30,620)
test virial_ewald ... bench: 133,107,701 ns/iter (+/- 5,069,337)
test virial_wolf  ... bench:     970,740 ns/iter (+/- 33,092)

test result: ok. 0 passed; 0 failed; 0 ignored; 6 measured
```

As we can see, there is something going on with ewald forces and virial calculation. The virial calculation should not take longer than the forces, and the forces should also be in the same region as the energy. Fixing this will be left to a future PR!

### CODE REVIEW

File: src/core/src/sys/interactions.rs [hunk 1]
```diff
@@ -278,6 +299,34 @@ mod test {
...
 
     #[test]
+    fn sorting_pairs() {
+        assert_eq!(sort_pair(Kind(0), Kind(1)), sort_pair(Kind(1), Kind(0)));
+        assert_eq!(sort_pair(Kind(125), Kind(0)), sort_pair(Kind(0), Kind(125)));
+    }
+
+    #[test]
+    fn sorting_angles() {
+        assert_eq!(sort_angle(Kind(1), Kind(0), Kind(2)), sort_angle(Kind(2), Kind(0), Kind(1)));
```
Comment (antoinewdg) [inline src/core/src/sys/interactions.rs]: In all these tests about sorting, we never check that a sequence is actually sorted.
Here both results could be 0, 2, 1 and the test would still pass.

**Comment by antoinewdg** [comment]
Well that's awesome !
A small remark: when you hide some bugfixes inside a PR for something else they are much harder to review :(

File: src/core/src/sys/interactions.rs (hunk 1 above)
Comment (Luthaf) [inline src/core/src/sys/interactions.rs]: Yes, naming is bad. This function is more about normalisation than sorting. We don't actually care that they are sorted, only that equivalent bonds/angles.dihedrals have an unique representation.

**Comment by Luthaf** [comment]
> A small remark: when you hide some bugfixes inside a PR for something else they are much harder to review :(

Sorry about that, I just went ahead and fixed the bugs on the way when implementing the propane benchmark. I'll try to have two PR next time.

File: src/core/src/sys/interactions.rs (hunk 1 above)
Comment (Luthaf) [inline src/core/src/sys/interactions.rs]: But we could check the normalisation explicitly, I'll add that

**Comment by Luthaf** [comment]
And the new results of the benchmarks with MC cache checking (slightly different from the first message, because on a different computer)

```
$ cargo bench
   Compiling lumol v0.1.0 (file:///Users/guillaume/code/science/lumol/lumol)
    Finished release [optimized] target(s) in 1.98 secs
     Running target/release/deps/argon-80320ec6ae378c54

running 4 tests
test cache_move_particle ... bench:     187,117 ns/iter (+/- 123,745)
test energy              ... bench:   2,238,396 ns/iter (+/- 445,629)
test forces              ... bench:   2,537,238 ns/iter (+/- 241,111)
test virial              ... bench:   2,919,467 ns/iter (+/- 457,942)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured

     Running target/release/deps/lumol-400725a8258366ee

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running target/release/deps/nacl-c72691cf22477026

running 8 tests
test cache_move_particle_ewald ... bench:     327,931 ns/iter (+/- 85,496)
test cache_move_particle_wolf  ... bench:      92,176 ns/iter (+/- 30,913)
test energy_ewald              ... bench:     571,534 ns/iter (+/- 110,843)
test energy_wolf               ... bench:     466,896 ns/iter (+/- 87,097)
test forces_ewald              ... bench:  37,146,759 ns/iter (+/- 3,829,054)
test forces_wolf               ... bench:     514,997 ns/iter (+/- 52,875)
test virial_ewald              ... bench:  92,760,161 ns/iter (+/- 10,252,266)
test virial_wolf               ... bench:     588,415 ns/iter (+/- 122,755)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured

     Running target/release/deps/propane-3f625718e17cdcfe

running 5 tests
test cache_move_all_rigid_molecules ... bench:   1,242,136 ns/iter (+/- 315,361)
test cache_move_particles           ... bench:     269,544 ns/iter (+/- 59,276)
test energy                         ... bench:   1,258,368 ns/iter (+/- 267,012)
test forces                         ... bench:   1,438,240 ns/iter (+/- 263,723)
test virial                         ... bench:   1,517,064 ns/iter (+/- 249,323)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured

     Running target/release/deps/water-00d22863dd85fd11

running 10 tests
test cache_move_all_rigid_molecules_ewald ... bench:   8,429,475 ns/iter (+/- 1,356,565)
test cache_move_all_rigid_molecules_wolf  ... bench:   8,313,306 ns/iter (+/- 1,893,437)
test cache_move_particles_ewald           ... bench:     787,949 ns/iter (+/- 202,144)
test cache_move_particles_wolf            ... bench:     428,027 ns/iter (+/- 126,250)
test energy_ewald                         ... bench:     982,367 ns/iter (+/- 210,628)
test energy_wolf                          ... bench:     711,763 ns/iter (+/- 152,194)
test forces_ewald                         ... bench:  53,170,056 ns/iter (+/- 4,049,052)
test forces_wolf                          ... bench:     909,617 ns/iter (+/- 147,866)
test virial_ewald                         ... bench: 145,000,724 ns/iter (+/- 11,442,685)
test virial_wolf                          ... bench:     892,583 ns/iter (+/- 186,399)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured
```

We have some low hanging fruits in the `cache_move_all_rigid_molecules` with coulomb interactions!

**Comment by Luthaf** [comment]
Yes, I am done here!

**Comment by Luthaf** [comment]
I did not see that you squash the whole PR here =)

Could you merge or rebase & merge the next time ? I like to preserve history and have smaller commits. Thanks!

**Comment by antoinewdg** [comment]
Yeah I'll do that next time, sorry.

**Pull request merged**